### PR TITLE
Turn on HAFS RTs on WCOSS2; AQM radians to degrees; NRL Neptune model 32-bit physics support

### LINKED ISSUE #1085: HAFS nested tests  failed on several platforms
## Description
The RT test hafs_regional_telescopic_2nests_atm failed a couple times on Orion when new baseline is created. The baseline was created successfully. But the RT test against the baseline failed with error: 

 Comparing atmf006.nc .........OK
 Comparing sfcf006.nc .........OK
 Comparing atm.nest02.f006.nc ............ALT CHECK......ERROR

It turned out that the compare_ncfile.py failed when comparing the atm.nest02.f006.nc from baseline and from the RT test:
 
compare_ncfile.py atm.nest02.f006.nc /work/noaa/stmp/bcurtis/stmp/bcurtis/FV3_RT/rt_410503/hafs_regional_telescopic_2nests_atm/atm.nest02.f006.nc
Traceback (most recent call last):
  File "/work/noaa/nems/emc.nemspara/autort/pr/867250832/20220304164615/ufs-weather-model/tests/compare_ncfile.py", line 14, in <module>
    if np.shape(nc1[varname][:])!=np.shape(nc2[varname][:]):
  File "netCDF4/_netCDF4.pyx", line 4408, in netCDF4._netCDF4.Variable.__getitem__
  File "netCDF4/_netCDF4.pyx", line 5352, in netCDF4._netCDF4.Variable._get
  File "netCDF4/_netCDF4.pyx", line 1887, in netCDF4._netCDF4._ensure_nc_success

When rerun the test without using baseline creation, there is no issue with file comparison.  

### To Reproduce:
on orion, check out model code:
cd ufs-weather-model/tests
./rt.sh -c -e
./rt.sh -m -e 

The orion log file show:
FAILED TESTS:
Test hafs_regional_telescopic_2nests_atm 104 failed in run_test failed

### LINKED ISSUE #1213: Bring in AQM changes to fix issue with PM2.5 peaks being offset
## Description
PM2.5 peaks are offset and there's a fix in AQM to change coordinate systems from radians to degrees that fixes that issue.

### LINKED ISSUE #1226: Turn on HAFS related regression tests on WCOSS2
## Description
Turn on HAFS RTs on WCOSS2 and switch to use the netcdf (instead of netcdf_parallel) write_grid_component option for HAFS RTs.
Note: when using the netcdf_parallel write grid component option, the forecast will hang on WCOSS2 during writing out netcdf history output files.

## Solution
Modify rt.conf and tests/hafs* files to turn on HAFS RTs on WCOSS2 and to use the netcdf write_grid_component option for HAFS RTs

## Related to
- addressed by ufs-community/ufs-weather-model/pull/1227

### LINKED ISSUE #1229: Update to use fms/2022.01 on wcoss_cray
## Description
On other ufs-weather-model supported platforms, the FMS version used has been updated to fms/2022.01. However, on wcoss_cray, it is still fms/2021.04, e.g.,
https://github.com/ufs-community/ufs-weather-model/blob/develop/modulefiles/ufs_wcoss_cray#L61
https://github.com/ufs-community/ufs-weather-model/blob/develop/modulefiles/ufs_wcoss_cray_debug#L61

## Solution
Update to use fms/2022.01 on wcoss_cray.

## Related to
Directly reference any issues or PRs in this or other repositories that this is related to, and describe how they are related. Example:
- addressed by ufs-community/ufs-weather-model/pull/1227

### PR DESCRIPTION
**Pull request by BinLiu-NOAA**
# PR Checklist

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [x] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.
  - Need at least create new baseline for the HAFS RTs on WCOSS2 
  - Need hafs_regional_storm_following_1nest_atm on hera.intel

- [ ] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsibility to keep the PR up-to-date with the develop branch of ufs-weather-model. 

## Description

Provide a detailed description of what this PR does. What bug does it fix, or what feature does it add? Is a change of answers expected from this PR? Are any library updates included in this PR (modulefiles etc.)?

 - Turn on HAFS RTs on WCOSS2 and switch to use the netcdf (instead of netcdf_parallel) write_grid_component option for HAFS RTs. 
   - When using the netcdf_parallel write grid component option, the forecast will hang on WCOSS2 during writing out netcdf history output files. There were similar issues on other platforms occasionally as well. 
 - Update to use fms/2022.01 on wcoss_cray.
 - This PR combines PR #1212
   - To fix an offset of PM2.5 peaks a coordinate change from radians to degrees was made.
   - There are no changes to results as there are no RT's for AQM yet.
   - Recommend bringing this PR in before AQM RT addition or along with it. 
 - This PR also combines PR #1206
   - Adds ccpp-physics changes needed to support 32-bit physics on Neptune. This does not add 32-bit physics to FV3, which requires quite a lot more changes in several repositories.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!) 
- fixes ufs-community/ufs-weather-model/issues/1226
- fixes ufs-community/ufs-weather-model/issues/1229
- potentially fix ufs-community/ufs-weather-model/issues/1085
- fixes ufs-community/ufs-weather-model/issues/1213
- Implements what this PR was supposed to:
  - https://github.com/NCAR/ccpp-physics/pull/797
- Which is about half of this batch of changes from NRL:
  - https://github.com/NCAR/ccpp-physics/pull/772

## Testing

Tests conducted on WCOSS2 run through all HAFS RTs.

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [x] wcoss2.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3
- [ ] opnReqTest for newly added/changed feature
- [x] CI (https://github.com/ufs-community/ufs-weather-model/actions/runs/2390284649)

## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).
- waiting on NOAA-EMC/AQM/pull/2
- waiting on NCAR/ccpp-physics/pull/918
- waiting on NOAA-EMC/fv3atm/pull/532

### CODE REVIEW

File: tests/rt.conf [hunk 1]
```diff
@@ -148,26 +148,26 @@ RUN     | control_wam_debug
...
+RUN     | hafs_regional_1nest_atm                                                                                                 | - jet.intel                             | fv3 |
+RUN     | hafs_regional_telescopic_2nests_atm                                                                                     | - jet.intel                             | fv3 |
+RUN     | hafs_global_1nest_atm                                                                                                   | - jet.intel                             | fv3 |
+RUN     | hafs_global_multiple_4nests_atm                                                                                         | - jet.intel wcoss_cray                  | fv3 |
+RUN     | hafs_regional_specified_moving_1nest_atm                                                                                | - jet.intel                             | fv3 |
+RUN     | hafs_regional_storm_following_1nest_atm                                                                                 | - jet.intel hera.intel                  | fv3 |
```
Comment (DeniseWorthen) [inline tests/rt.conf]: Please turn this test back on for hera.intel

**Comment by BrianCurtis-NOAA** [comment]
@BinLiu-NOAA Please add a WCOSS2 item in the ufs-weather-model/.github/pull_request_template.md

File: tests/rt.conf (hunk 1 above)
Comment (BinLiu-NOAA) [reply tests/rt.conf]: Done. Thanks, @DeniseWorthen!

**Comment by BinLiu-NOAA** [comment]
> @BinLiu-NOAA Please add a WCOSS2 item in the ufs-weather-model/.github/pull_request_template.md

@BrianCurtis-NOAA, done. Thanks!

**Comment by BrianCurtis-NOAA** [comment]
@BinLiu-NOAA Its now your turn at the top of the commit queue. Please update .gitmodules with #1206 (fv3atm) changes and #1212 (AQM) changes.

**Comment by BrianCurtis-NOAA** [comment]
Also merge with develop branch if you haven't yet. Nevermind, looks like you did.

**Comment by BinLiu-NOAA** [comment]
> @BinLiu-NOAA Its now your turn at the top of the commit queue. Please update .gitmodules with #1206 (fv3atm) changes and #1212 (AQM) changes.

@BrianCurtis-NOAA, Done! Please check with @SamuelTrahanNOAA and @JianpingHuang-NOAA whether or not the submodules are ready for regression tests. Thanks!

**Comment by BrianCurtis-NOAA** [comment]
@BinLiu-NOAA Please add me (BrianCurtis-NOAA) as a collaborator on your hafs-community/ufs-weather-model fork. (i didn't see it last time I checked)

**Comment by BinLiu-NOAA** [comment]
> @BinLiu-NOAA Please add me (BrianCurtis-NOAA) as a collaborator on your hafs-community/ufs-weather-model fork. (i didn't see it last time I checked)
@BrianCurtis-NOAA, Just checked, you were already there. Let me know otherwise.

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: cheyenne
Compiler: gnu
Job: RT
[RT] Repo location: /glade/scratch/epicufsrt/autort/tests/auto/pr/942703315/20220525144512/ufs-weather-model
[RT] Error: Test compile_003 failed in run_compile failed 
Please make changes and add the following label back: cheyenne-gnu-RT

**Comment by BrianCurtis-NOAA** [comment]
@SamuelTrahanNOAA can you look at /glade/scratch/epicufsrt/FV3_RT/rt_35192/compile_003 on Cheyenne?

**Comment by DeniseWorthen** [comment]
The compile_003 failure seems to be because of non-typing of variables, eg:

```
/glade/scratch/epicufsrt/autort/tests/auto/pr/942703315/20220525144512/ufs-weather-model/FV3/ccpp/physics/physics/module_BL_MYJPBL.F90:1304:11:

 1304 |       (NTSD,ME,LMH,LMXL,GM,GH,EL,T,Q2,Z,AKM,AKH,I,J,LM,PRINT_DIAG,KPBL)
      |           1
Error: Symbol ‘ntsd’ at (1) has no IMPLICIT type
```

**Comment by SamuelTrahanNOAA** [comment]
That bug comes from my PR. You need to merge my latest fix in the ccpp-physics repo. (Sorry, I forgot to push the bug fix.)

**Comment by BrianCurtis-NOAA** [comment]
> That bug comes from my PR. You need to merge my latest fix in the ccpp-physics repo. (Sorry, I forgot to push the bug fix.)

@BinLiu-NOAA can you update the hash for FV3?

**Comment by BrianCurtis-NOAA** [comment]
@BinLiu-NOAA can you look at /gpfs/hps3/stmp/Brian.Curtis/FV3_RT/rt_20138/hafs_regional_atm_thompson_gfdlsf ?

It fails and I've browsed it over but can't find out why.

**Comment by BinLiu-NOAA** [comment]
> > That bug comes from my PR. You need to merge my latest fix in the ccpp-physics repo. (Sorry, I forgot to push the bug fix.)
> 
> @BinLiu-NOAA can you update the hash for FV3?

Done by syncing @SamuelTrahanNOAA's ccpp-neptune branch. BTW, if I did not see the conversation in time, both of you (@BrianCurtis-NOAA and @SamuelTrahanNOAA) should already have written access to this PR branch (feature/hafs_wcoss2_rts). Please feel free to commit/push into this branch. Thanks!

**Comment by BinLiu-NOAA** [comment]
> @BinLiu-NOAA can you look at /gpfs/hps3/stmp/Brian.Curtis/FV3_RT/rt_20138/hafs_regional_atm_thompson_gfdlsf ?
> 
> It fails and I've browsed it over but can't find out why.

@BrianCurtis-NOAA, looks like we encountered the issue complaining "Variables are different" when comparing test output against baseline. See in compare_ncfile.log. So, most likely, either the baseline or this RT got corrupted netcdf output files.

**Comment by BrianCurtis-NOAA** [comment]
> > @BinLiu-NOAA can you look at /gpfs/hps3/stmp/Brian.Curtis/FV3_RT/rt_20138/hafs_regional_atm_thompson_gfdlsf ?
> > It fails and I've browsed it over but can't find out why.
> 
> @BrianCurtis-NOAA, looks like we encountered the issue complaining "Variables are different" when comparing test output against baseline. See in compare_ncfile.log. So, most likely, either the baseline or this RT got corrupted netcdf output files.

OK I will create that baseline again.

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: cheyenne
Compiler: intel
Job: RT
[RT] Repo location: /glade/scratch/epicufsrt/autort/tests/auto/pr/942703315/20220525143011/ufs-weather-model
[RT] Error: Test hafs_regional_atm_thompson_gfdlsf 096 failed in check_result failed 
[RT] Error: Test hafs_regional_atm_thompson_gfdlsf 096 failed in run_test failed 
Please make changes and add the following label back: cheyenne-intel-RT

**Comment by BrianCurtis-NOAA** [comment]
> _**Automated RT Failure Notification**_ Machine: cheyenne Compiler: intel Job: RT [RT] Repo location: /glade/scratch/epicufsrt/autort/tests/auto/pr/942703315/20220525143011/ufs-weather-model [RT] Error: Test hafs_regional_atm_thompson_gfdlsf 096 failed in check_result failed [RT] Error: Test hafs_regional_atm_thompson_gfdlsf 096 failed in run_test failed Please make changes and add the following label back: cheyenne-intel-RT

I'm going to guess that we need new baselines for that test on all systems? It failed on Dell/Cray and Cheyenne.intel

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: hera
Compiler: gnu
Job: RT
[RT] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/942703315/20220525204519/ufs-weather-model
[RT] Error: Test compile_003 failed in run_compile failed 
Please make changes and add the following label back: hera-gnu-RT

**Comment by BrianCurtis-NOAA** [comment]
> _**Automated RT Failure Notification**_ Machine: hera Compiler: gnu Job: RT [RT] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/942703315/20220525204519/ufs-weather-model [RT] Error: Test compile_003 failed in run_compile failed Please make changes and add the following label back: hera-gnu-RT

This was known, will re-run with intel once the baseline is moved over

**Comment by BrianCurtis-NOAA** [comment]
@BinLiu-NOAA With the error of variables being different in the hafs_regional_.._gfdl test, you're OK with regenerating that one test, correct?

**Comment by BinLiu-NOAA** [comment]
> @BinLiu-NOAA With the error of variables being different in the hafs_regional_.._gfdl test, you're OK with regenerating that one test, correct?

Yes, please help to go ahead regenerating baseline for this test. There is a change from netcdf_parallel to netcdf output option, which might have affected the netcdf output files. Thanks!

**Comment by junwang-noaa** [comment]
@BinLiu-NOAA Would you please check the differences between the two files(netcdf_parallel and netcdf) to confirm the data fields are correct? Thanks

**Comment by BinLiu-NOAA** [comment]
> @BinLiu-NOAA Would you please check the differences between the two files(netcdf_parallel and netcdf) to confirm the data fields are correct? Thanks

@junwang-noaa and @BrianCurtis-NOAA, Sure, doing that. Actually found an interesting issue when comparing the new generated baseline:
ncdump -h /gpfs/hps3/emc/nems/noscrub/emc.nemspara/RT/NEMSfv3gfs/develop-20220519/hafs_regional_atm_thompson_gfdlsf/atmf006.nc
against the previous version baseline on wcoss_cray:
ncdump -h /gpfs/hps3/emc/nems/noscrub/emc.nemspara/RT/NEMSfv3gfs/develop-20220516/hafs_regional_atm_thompson_gfdlsf/atmf006.nc

The new baseline got the following three variables missing, for which I don't know why. Other than that the other variables actually got identical values. I believe the only thing this PR changed for this hafs_regional_atm_thompson_gfdlsf test is using netcdf instead of netcdf_parallel output. Also, another special thing for this test is it uses the "cubed_sphere" native grid output option.

        char time_iso(time, nchars) ;
                time_iso:long_name = "valid time" ;
                time_iso:description = "ISO 8601 datetime string" ;
        double lon(grid_yt, grid_xt) ;
                lon:long_name = "T-cell longitude" ;
                lon:units = "degrees_E" ;
        double lat(grid_yt, grid_xt) ;
                lat:long_name = "T-cell latitude" ;
                lat:units = "degrees_N" ;

**Comment by SamuelTrahanNOAA** [comment]
The fv3gfs_diag_output never sends those three variables directly, and they're not in the diag_table. Maybe the diag manager is supposed to add them automatically? If so, this is a bug in FMS, beyond Bin's control.

*Edit: They're not in any diag tables in tests/parm/diag_table*

**Comment by DusanJovic-NOAA** [comment]
> > @BinLiu-NOAA Would you please check the differences between the two files(netcdf_parallel and netcdf) to confirm the data fields are correct? Thanks
> 
> @junwang-noaa and @BrianCurtis-NOAA, Sure, doing that. Actually found an interesting issue when comparing the new generated baseline: ncdump -h /gpfs/hps3/emc/nems/noscrub/emc.nemspara/RT/NEMSfv3gfs/develop-20220519/hafs_regional_atm_thompson_gfdlsf/atmf006.nc against the previous version baseline on wcoss_cray: ncdump -h /gpfs/hps3/emc/nems/noscrub/emc.nemspara/RT/NEMSfv3gfs/develop-20220516/hafs_regional_atm_thompson_gfdlsf/atmf006.nc
> 
> The new baseline got the following three variables missing, for which I don't know why. Other than that the other variables actually got identical values. I believe the only thing this PR changed for this hafs_regional_atm_thompson_gfdlsf test is using netcdf instead of netcdf_parallel output. Also, another special thing for this test is it uses the "cubed_sphere" native grid output option.
> 
> ```
>     char time_iso(time, nchars) ;
>             time_iso:long_name = "valid time" ;
>             time_iso:description = "ISO 8601 datetime string" ;
>     double lon(grid_yt, grid_xt) ;
>             lon:long_name = "T-cell longitude" ;
>             lon:units = "degrees_E" ;
>     double lat(grid_yt, grid_xt) ;
>             lat:long_name = "T-cell latitude" ;
>             lat:units = "degrees_N" ;
> ```

Different output routines are used for 'cubed_sphere_grid' output_grid with 'netcdf' (ESMFproto_FieldBundleWrite) than for 'netcdf_parallel' (write_netcdf).

**Comment by junwang-noaa** [comment]
@BinLiu-NOAA Thanks for checking. I think the valid time was only added to non-native grid history file output. We are using ESMF calls to write out native grid output. This is a different issue not related to this PR. @DusanJovic-NOAA FYI.

**Comment by BinLiu-NOAA** [comment]
> Different output routines are used for 'cubed_sphere_grid' output_grid with 'netcdf' (ESMFproto_FieldBundleWrite) than for 'netcdf_parallel' (write_netcdf).

Thanks for this information, @DusanJovic-NOAA! With that, @junwang-noaa and @BrianCurtis-NOAA, I would suggest just moving forward with the new re-generated baseline for this specific hafs_regional_atm_thompson_gfdlsf test. And let this issue/inconsistency be fixed/addressed in a future PR. Thanks!

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: gaea
Compiler: intel
Job: RT
[RT] Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/942703315/20220526113015/ufs-weather-model
[RT] Error: Test hafs_regional_atm_thompson_gfdlsf 095 failed in check_result failed 
[RT] Error: Test hafs_regional_atm_thompson_gfdlsf 095 failed in run_test failed 
Please make changes and add the following label back: gaea-intel-RT

**Comment by junwang-noaa** [comment]
@BrianCurtis-NOAA Are you going to rerun the hafs_regional_atm_thompson_gfdlsf test?

**Comment by BrianCurtis-NOAA** [comment]
> @BrianCurtis-NOAA Are you going to rerun the hafs_regional_atm_thompson_gfdlsf test?

Yes I am. @DeniseWorthen did a compare and found new/previous baseline to be the same even though it showed as failed. I started the re-run before 3 but it's stuck in the Gaea queue still. Apologies it's going slow today.

**Comment by BrianCurtis-NOAA** [comment]
@junwang-noaa ready for merges for ccpp-physics and AQM

**Comment by BrianCurtis-NOAA** [comment]
@BinLiu-NOAA Please update .gitmodules and AQM [73da18b](https://github.com/NOAA-EMC/AQM/commit/73da18b422c2b2dcf1e15a1d21fbba12a0479059) and FV3 [515a39d](https://github.com/NOAA-EMC/fv3atm/commit/515a39d5f43c297d7d56546fb92289e9b653f7f4) hashes

**Review by DusanJovic-NOAA — APPROVED** [review]

**Review by MinsukJi-NOAA — APPROVED** [review]

**Pull request merged**